Migrate/Outcome-Service

sodabottlebaby opened the pull request:
# Description

The component `outcome-service` is a service that calculates and manages learning outcomes based on the student's target grades and contribution. It enables users to select a task's alignment to each learning outcome (LO) ranging across six labels and creates a map.

## Type of change

- [x] Bug fix 
- [x] This change requires a documentation update

## How Has This Been Tested?

Login using student_1 and navigate to a unit with learning outcomes. Select from 1 - 5 rating values, and add a provided rationale (optional). Navigate to outcome achievement noting progress.

before:
<img width="1502" alt="OutcomeServiceAlignments_Before" src="https://github.com/user-attachments/assets/8a886a69-b6b6-4a7b-80e0-6c4504744bcf" />
<img width="1502" alt="OutcomeServiceAchievement_Before" src="https://github.com/user-attachments/assets/5872f289-f9d4-483b-8244-f41e1cda24ad" />

after:
<img width="1502" alt="OutcomeServiceAlignment_After" src="https://github.com/user-attachments/assets/7784fdd7-1fcf-4b1d-bed7-ae83241be60d" />
<img width="1502" alt="OutcomeServiceAchievement_After" src="https://github.com/user-attachments/assets/080c65f9-ffb2-463b-9ca3-c4cf8b99e4e2" />




- [x] Tested in latest Chrome
- [x] Tested in latest Safari
- [x] Tested in latest Firefox

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

## Folders and Files Added/Modified

- Added:
  - [x] outcome-service.ts
- Modified:
  - [x] outcome-service.coffee

sodabottlebaby committed:
fix(services): migrated outcome-service.coffee to outcome-service.ts

returnMarcco commented:
@sodabottlebaby Hello. Good job on your first PR. There are a few things that require changing.

1. Your work was branched off the `development` branch - you need to do this work again by branching off the `9.x` branch as stated here in the ticket. https://teams.microsoft.com/l/entity/com.microsoft.teamspace.tab.planner/tt.c_19:bd20175d09414f079490a2403f7fca74@thread.tacv2_p_njykIFLDn0iAY1at7tACfcgADgBS_h_1714385623719?tenantId=d02378ec-1688-46d5-8540-1c28b5f470f6&webUrl=https%3A%2F%2Ftasks.teams.microsoft.com%2Fteamsui%2FpersonalApp%2Falltasklists&context=%7B%22subEntityId%22%3A%22%2Fv1%2Fplan%2FnjykIFLDn0iAY1at7tACfcgADgBS%2Ftask%2FVsrHKLE8rUmmGdH0scoYDsgAKspo%22%2C%22channelId%22%3A%2219%3Abd20175d09414f079490a2403f7fca74%40thread.tacv2%22%7D

    - Note:  When choosing the branch to `merge into`, choose the `thoth-tech:9.x` branch. For the `from` branch, it should be the `sodabottlebaby:9.x` branch.
![Screenshot 2025-04-10 204248](https://github.com/user-attachments/assets/0b1ab634-5f63-4a60-af80-fbd40d2a964c)

    - Note: If you are using `Windows`and have trouble building the `9.x` branch, please see this thread: https://teams.microsoft.com/l/message/19:bd20175d09414f079490a2403f7fca74@thread.tacv2/1743634833688?tenantId=d02378ec-1688-46d5-8540-1c28b5f470f6&groupId=0e15669c-3f66-49aa-b023-640fe1dda2e0&parentMessageId=1743634833688&teamName=Thoth%20Tech&channelName=OnTrack&createdTime=1743634833688

2. Please `unlink` the old service from the app, and `link` your new service. Read here for instructions on how to do this: https://github.com/thoth-tech/doubtfire-web/blob/development/MIGRATION-GUIDE.md#step-3---unlink-old-component-and-add-in-new

3. Please write a `unit test` for the new service. A unit test is an automated test for frontend code, and checks that the functions you write produce the expected output, for the supplied input. You can use the `/common/services/grade.service.spec.ts` file as a guide on how to do this.

Once these changes have been made, commited, and pushed to origin, please re-request another peer review from me and I'll review the actual Typescript code.

Cheers,
Jason

amriith commented:
Hi @sodabottlebaby  Your PR looks good overall — great work! However, there are a few changes that need to be addressed to ensure the migration works correctly:  
1. In doubtfire-angularjs.module.ts, the OutcomeService should be imported like this `import { OutcomeService } from 'src/app/common/services/outcome-service'; `
2. The old CoffeeScript imports need to be fully replaced with their TypeScript equivalents.
3. The integration with LearningOutcomeService still needs to be completed.
4. A few interface definitions seem to be missing — adding them will ensure type safety and consistency.
 Let me know if you need any clarification on these points. Happy to help.

sodabottlebaby committed:
fix(services): added unit test for outcome-service.ts and change imports

- Made changes according to review comments
- Included integration with LearningOutcomeService
- Changed branch from development to 9.x
- Unlinked old component and added in new
- Added a unit test

returnMarcco reviewed (approved):
Please see my review comment in this discussion. Looks good to me.
